[Merged by Bors] - refactor(FunctionField): use `exp` instead of `ofAdd`

Louddy opened the pull request:
Make the valuation at infinity in `FunctionField` use the notation `WithZero.exp` instead of `Multiplicative.ofAdd`.

I have also tagged many of the `exp` lemmas with simp making most basic results provable with simp.
I have left the proofs with `rw` in this file as this is faster, but having the option to close similar goals with simp is nice for further development.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

YaelDillies commented on Mathlib/Algebra/GroupWithZero/WithZero.lean:
```diff
@@ -402,6 +408,7 @@ def expEquiv : G ≃ (Gᵐ⁰)ˣ := Multiplicative.ofAdd.trans unitsWithZeroEqui
 def logEquiv : (Gᵐ⁰)ˣ ≃ G := unitsWithZeroEquiv.toEquiv.trans Multiplicative.toAdd
 
 @[simp] lemma logEquiv_symm : (logEquiv (G := G)).symm = expEquiv := rfl
+
 @[simp] lemma expEquiv_symm : (expEquiv (G := G)).symm = logEquiv := rfl
```
I personally like having pairs of one-liners lined up, to show they are the same

YaelDillies commented on Mathlib/Algebra/GroupWithZero/WithZero.lean:
```diff
@@ -415,16 +422,25 @@ def logEquiv : (Gᵐ⁰)ˣ ≃ G := unitsWithZeroEquiv.toEquiv.trans Multiplicat
 
 lemma logEquiv_unitsMk0 (x : Gᵐ⁰) (hx) : logEquiv (.mk0 x hx) = log x := logEquiv_apply _
 
-lemma exp_sub (a b : G) : exp (a - b) = exp a / exp b := rfl
+@[simp]
+lemma exp_div_exp (a b : G) : exp a / exp b = exp (a - b) := rfl
```
Why turn this one around? You made `exp (a + b) = exp a * exp b` simp earlier

Louddy replied:
(This reasoning comes from an application that I have.)

Louddy replied:
I agree, I removed `exp_add` as simp and added `exp_mul_exp`.

Louddy replied:
Maybe I am extrapolating too much but this comes from `inftyValuation.X_zpow` in this PR. I want the exponent to get absorbed and this is why I changed the direction to have `zpow_exp`. See [here](https://live.lean-lang.org/#codez=JYWwDg9gTgLgBAWQIYwBYBtgCMBQOJgCmAdnAOrBoBahUEehAHkuOoXABQhwBccgJIQBKXpyZg4ARmEA9ONwC8oxuJDCeirAE8ccOAGdQ4gNqAEwjhiA+gC89IAK7oAukA) for example.

YaelDillies replied:
`zpow_exp` is *not* a distributivity lemma, contrary to `exp_add`/`exp_sub`, therefore it is much easier to agree to changing its direction

YaelDillies commented on Mathlib/NumberTheory/FunctionField.lean:
```diff
@@ -196,23 +195,27 @@ def inftyValuation : Valuation (RatFunc Fq) ℤᵐ⁰ where
 theorem inftyValuation.C {k : Fq} (hk : k ≠ 0) :
-    inftyValuationDef Fq (RatFunc.C k) = Multiplicative.ofAdd (0 : ℤ) := by
-  simp [hk]
+    inftyValuation Fq (RatFunc.C k) = exp 0 := by
```
```suggestion
    inftyValuation Fq (RatFunc.C k) = 1 := by
```

YaelDillies reviewed (commented):
Looks very nice! but I am confused by the lemmas you tag with simp

pechersky commented on Mathlib/NumberTheory/FunctionField.lean:
```diff
@@ -162,8 +164,7 @@ theorem InftyValuation.map_mul' (x y : RatFunc Fq) :
     · rw [hy, mul_zero, if_pos (Eq.refl _), mul_zero]
-    · rw [if_neg hx, if_neg hy, if_neg (mul_ne_zero hx hy), ← WithZero.coe_mul, WithZero.coe_inj,
-        ← ofAdd_add, RatFunc.intDegree_mul hx hy]
+    · simp_all [ RatFunc.intDegree_mul, exp_add]
```
```suggestion
    · simp_all [RatFunc.intDegree_mul]
```
spacing and `exp_add` is now simp tagged

pechersky commented on Mathlib/NumberTheory/FunctionField.lean:
```diff
@@ -141,19 +141,21 @@ end ringOfIntegers
+`exp (degree(f) - degree(g))`. -/
 def inftyValuationDef (r : RatFunc Fq) : ℤᵐ⁰ :=
-  if r = 0 then 0 else ↑(Multiplicative.ofAdd r.intDegree)
+  if r = 0 then 0 else exp r.intDegree
```
@YaelDillies this `if 0 then 0 else exp` pattern again

pechersky commented on Mathlib/NumberTheory/FunctionField.lean:
```diff
@@ -196,23 +195,27 @@ def inftyValuation : Valuation (RatFunc Fq) ℤᵐ⁰ where
 theorem inftyValuation.C {k : Fq} (hk : k ≠ 0) :
-    inftyValuationDef Fq (RatFunc.C k) = Multiplicative.ofAdd (0 : ℤ) := by
-  simp [hk]
+    inftyValuation Fq (RatFunc.C k) = exp 0 := by
+  simp [inftyValuation_apply, hk]
```
shouldn't this be `= 1` directly?

pechersky commented on Mathlib/NumberTheory/FunctionField.lean:
```diff
@@ -221,7 +224,7 @@ def inftyValuedFqt : Valued (RatFunc Fq) ℤᵐ⁰ :=
 
 theorem inftyValuedFqt.def {x : RatFunc Fq} :
-    @Valued.v (RatFunc Fq) _ _ _ (inftyValuedFqt Fq) x = inftyValuationDef Fq x :=
+    Valued.v (self := inftyValuedFqt Fq) x = inftyValuationDef Fq x :=
```
My preference would be
```suggestion
    (inftyValuedFqt Fq).v x = inftyValuationDef Fq x :=
```

pechersky replied:
and similarly below for the completion instance and lemmas

pechersky commented on Mathlib/NumberTheory/Padics/PadicNumbers.lean:
```diff
@@ -1071,7 +1071,7 @@ noncomputable def mulValuation : Valuation ℚ_[p] ℤᵐ⁰ where
   map_zero' := by simp
   map_one' := by simp
-  map_mul' _ := by split_ifs <;> simp_all [exp_add, exp_neg, mul_comm]
+  map_mul' _ := by split_ifs <;> simp_all [exp_add, mul_comm]
```
```suggestion
  map_mul' _ := by split_ifs <;> simp_all [mul_comm]
```

pechersky reviewed (commented):
Thanks! Seems similar to #27339.

faenuccio commented:
> Thanks! Seems similar to #27339.

Indeed, is it possible to coordinate with @Louddy so that we don't have duplicates and we see clearly what is done where?

Louddy commented:
I have added `exp_add`/`exp_sub` as simp lemmas for coherence with other simp lemmas like `map_add`. So that simp reduces obvious terms like `exp(x) * exp (-x)`, I have also added `exp_neg` as a simp lemma. Having these in the simp set creates problems in two proofs where I need to hide `exp_neg`.

Louddy commented on Mathlib/NumberTheory/FunctionField.lean:
```diff
@@ -196,23 +195,27 @@ def inftyValuation : Valuation (RatFunc Fq) ℤᵐ⁰ where
   map_add_le_max' := InftyValuation.map_add_le_max' Fq
 
-@[simp]
 theorem inftyValuation_apply {x : RatFunc Fq} : inftyValuation Fq x = inftyValuationDef Fq x :=
```
Unfortunately not, it seems to me that the objective was to pull everything out of `inftyValuation`, including the data.

Louddy commented on Mathlib/NumberTheory/Padics/PadicNumbers.lean:
```diff
@@ -1071,10 +1071,10 @@ noncomputable def mulValuation : Valuation ℚ_[p] ℤᵐ⁰ where
   map_add_le_max' _ _ := by
     split_ifs
-    any_goals simp_all
+    any_goals simp_all [-exp_neg]
```
` (exp (x + y).valuation)⁻¹ ≤ (exp x.valuation)⁻¹ ∨ (exp (x + y).valuation)⁻¹ ≤ (exp y.valuation)⁻¹` which prevents `exp_le_exp` from triggering.

YaelDillies replied:
Then
```suggestion
    any_goals simp_all [inv_le_inv₀]
```
should work :-)

Louddy commented on Mathlib/NumberTheory/Padics/ValuativeRel.lean:
```diff
@@ -33,7 +33,8 @@ lemma valuation_p_ne_zero : v p ≠ 0 := by
 @[simp]
 lemma valuation_p_lt_one : v p < 1 := by
-  simp [(isEquiv v (Padic.mulValuation)).lt_one_iff_lt_one, hp.out.ne_zero, ← lt_log_iff_exp_lt]
+  simp [(isEquiv v (Padic.mulValuation)).lt_one_iff_lt_one, hp.out.ne_zero,
+    - exp_neg, ← lt_log_iff_exp_lt]
```
`(exp 1)⁻¹ < 1`, we can maybe add a simp lemma than handles this?

YaelDillies reviewed (approved):
Thanks! 🚀

maintainer merge
